File: src/main/scala/com/exasol/cloudetl/filesystem/FileSystemManager.scala
package com.exasol.cloudetl.filesystem

import java.io.FileNotFoundException
import java.nio.file.Path

import com.exasol.errorreporting.ExaError

import org.apache.hadoop.fs.{Path => HadoopPath}
import org.apache.hadoop.fs.FileStatus
import org.apache.hadoop.fs.FileSystem

/**
 * A filesystem helper class.
 */
final case class FileSystemManager(fileSystem: FileSystem) {

  /**
   * Returns list of files from path with or without globbing.
   *
   * It also adds paths from the sub-directories to the returned list.
   *
   * @param path input path
   * @return list of paths
   */
  def getFiles(path: String): Seq[HadoopPath] = globStatus(path)

  /**
   * Returns list of files from local path.
   */
  def getLocalFiles(path: Path): Seq[HadoopPath] =
    getFiles(path.toAbsolutePath().toUri().getRawPath())

  private[this] def globStatus(path: String): Seq[HadoopPath] = {
    val hadoopPath = new HadoopPath(path)
    val statuses = fileSystem.globStatus(hadoopPath, HiddenFilesFilter)
    if (statuses == null) {
      throw new FileNotFoundException(
        ExaError
          .messageBuilder("E-CSE-1")
          .message("Provided file path {{PATH}} does not exist.")
          .parameter("PATH", path)
          .mitigation("Please use valid file path.")
          .toString()
      )
    } else {
      if (isSingleDirectory(statuses)) {
        // User path is a single directory without any glob, only then list the files.
        listFiles(fileSystem.listStatus(statuses(0).getPath(), HiddenFilesFilter))
      } else {
        listFiles(statuses)
      }
    }
  }

  // Check if user provided path returns a single directory
  private[this] def isSingleDirectory(paths: Array[FileStatus]): Boolean =
    paths.length == 1 && fileSystem.getFileStatus(paths(0).getPath()).isDirectory()

  private[this] def listFiles(paths: Array[FileStatus]): Seq[HadoopPath] =
    paths.toSeq.filter(_.isFile()).map(_.getPath())

}

File: src/main/java/com/exasol/cloudetl/filesystem/FileSystemManager.java
package com.exasol.cloudetl.filesystem;

import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;

import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.errorreporting.ExaError;

/** Filesystem helper. */
public final class FileSystemManager {
    private final FileSystem fileSystem;

    /** Create a manager. */
    public FileSystemManager(final FileSystem fileSystem) {
        this.fileSystem = fileSystem;
    }

    /** Factory for Java callers. */
    public static FileSystemManager apply(final FileSystem fileSystem) {
        return new FileSystemManager(fileSystem);
    }

    /** Return matching files. */
    public scala.collection.immutable.Seq<Path> getFiles(final String path) throws java.io.IOException {
        return globStatus(path);
    }

    /** Return matching local files. */
    public scala.collection.immutable.Seq<Path> getLocalFiles(final java.nio.file.Path path) throws java.io.IOException {
        return getFiles(path.toAbsolutePath().toUri().getRawPath());
    }

    private scala.collection.immutable.Seq<Path> globStatus(final String path) throws java.io.IOException {
        final Path hadoopPath = new Path(path);
        final FileStatus[] statuses = this.fileSystem.globStatus(hadoopPath, HiddenFilesFilter.INSTANCE);
        if (statuses == null) {
            throw new FileNotFoundException(ExaError.messageBuilder("E-CSE-1")
                    .message("Provided file path {{PATH}} does not exist.", path).mitigation("Please use valid file path.")
                    .toString());
        }
        if (isSingleDirectory(statuses)) {
            return listFiles(this.fileSystem.listStatus(statuses[0].getPath(), HiddenFilesFilter.INSTANCE));
        }
        return listFiles(statuses);
    }

    private boolean isSingleDirectory(final FileStatus[] paths) throws java.io.IOException {
        return paths.length == 1 && this.fileSystem.getFileStatus(paths[0].getPath()).isDirectory();
    }

    private scala.collection.immutable.Seq<Path> listFiles(final FileStatus[] paths) {
        final List<Path> result = new ArrayList<>();
        for (final FileStatus path : paths) {
            if (path.isFile()) {
                result.add(path.getPath());
            }
        }
        return ScalaConverters.seqFromJava(result);
    }
}
